src/cluster_slot.c

```diff
@@ -55,53 +122,68 @@ static int slotStatEntryDescCmp(const void *a, const void *b) {
+            sorted[slot].stat = getSingleSlotStat(slot, order_by);
+        } else {
+            /* Even if the slot does not belong to the current shard,
+             * we should fill the entry with default values so that qsort() does not segfault.
+             * These entries will be filtered and ignored upon addReplySortedSlotStats(). */
+            sorted[slot].slot = DEFAULT_SLOT;
+            sorted[slot].stat = DEFAULT_STAT;
         }
     }
-    return C_OK;
+    qsort(sorted, CLUSTER_SLOTS, sizeof(sortedSlotStatEntry), (desc) ? slotStatEntryDescCmp : slotStatEntryAscCmp);
 }
 
-void addReplySlotStats(client *c, unsigned char *slots) {
-    int count = countValidSlots(slots);
-    addReplyMapLen(c, count);
+static void addReplySingleSlotStat(client *c, int slot) {
+    addReplyLongLong(c, slot);
+    addReplyMapLen(c, 1);
+    addReplyBulkCString(c, "key_count");
+    addReplyLongLong(c, (*server.db->slots_to_keys).by_slot[slot].count);
+}
+
+static void addReplySlotStats(client *c, unsigned char *slots) {
+    int num_slots_assigned = countAssignedSlotsFromSlotsArray(slots);
+    addReplyMapLen(c, num_slots_assigned);
 
     for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
         if (!slots[slot]) continue;
         addReplySingleSlotStat(c, slot);
     }
 }
 
-int checkSlotAssignmentsOrReply(client *c, unsigned char *slots, int del, int start_slot, int end_slot) {
-    for (int slot = start_slot; slot <= end_slot; slot++) {
-        if (del && server.cluster->slots[slot] == NULL) {
-            addReplyErrorFormat(c,"Slot %d is already unassigned", slot);
-            return C_ERR;
-        } else if (!del && server.cluster->slots[slot]) {
-            addReplyErrorFormat(c,"Slot %d is already busy", slot);
-            return C_ERR;
-        }
-        if (slots[slot]++ == 1) {
-            addReplyErrorFormat(c,"Slot %d specified multiple times",(int)slot);
-            return C_ERR;
-        }
+static void addReplySortedSlotStats(client *c, sortedSlotStatEntry sorted[], int limit) {
+    int num_slots_assigned = countAssignedSlotsFromEntireShard();
+    int len = min(limit, num_slots_assigned);
+    addReplyMapLen(c, len);
+
+    for (int i = 0; i < CLUSTER_SLOTS; i++) {
+        if (sorted[i].slot == DEFAULT_SLOT) continue;
+        if (len <= 0) break;
+
+        addReplySingleSlotStat(c, sorted[i].slot);
+        len--;
     }
-    return C_OK;
 }
 
-int checkSlotStatsOrderByArgumentOrReply(client *c, int *order_by, int *limit, int *desc) {
+static void sortAndAddReplySlotStats(client *c, int order_by, int limit, int desc) {
+    sortedSlotStatEntry sorted[CLUSTER_SLOTS];
+    sortSlotStats(sorted, order_by, desc);
+    addReplySortedSlotStats(c, sorted, limit);
+}
+
+static int checkSlotStatsOrderByArgumentOrReply(client *c, int *order_by, int *limit, int *desc) {
     *limit = CLUSTER_SLOTS;
     *desc = 1;
     *order_by = ORDER_BY_INVALID;
 
     if (!strcasecmp(c->argv[3]->ptr, "key_count")) {
```

Comment:
Need to check for arg size before accessing `c->argv[3]`. Currently, running `cluster slot-stats orderby` will crash the server.